### PR TITLE
Grid ufunc refactor

### LINKED ISSUE #344: Proposal: grid_ufunc
Xarray and the whole python array ecosystem has evolved considerably since the creation of xgcm. The introduction of xarray's `apply_ufunc`, dask / numba gufuncs, etc. has evolved the way we think about array code, allowing us to do complex array operations in a compact, elegant way. At the same time, we have become more aware of some of the performance limitations of xgcm, particularly with large dask arrays. This proposal is for an internal refactoring of xgcm that could address all of the above.

### Background: generatlized ufuncs

The numpy C api defines the concept of [generalized ufuncs](https://numpy.org/doc/stable/reference/c-api/generalized-ufuncs.html) (gufuncs). Dask [exposes gufuncs at the python level](https://docs.dask.org/en/latest/array-gufunc.html). Numba also allows you to write [vectorized gufuncs](https://numba.pydata.org/numba-doc/latest/user/vectorize.html#the-guvectorize-decorator).  At the core of gufuncs is the concept of a _signature_. A signature is a symbolic representation of how a function acts on a set of _core dimensions_. The specifies the required core dimensions of the inputs and how the output dimensions are related to the input dimensions. It's easiest to explain signatures through examples. Here I copy directly from the numpy docs linked above.

name | signature | common usage
-- | -- | --
add | `(),()->()` | binary ufunc
sum1d | `(i)->()` | reduction
inner1d | `(i),(i)->()` | vector-vector multiplication
matmat | `(m,n),(n,p)->(m,p)` | matrix multiplication
vecmat | `(n),(n,p)->(p)` | vector-matrix multiplication
matvec | `(m,n),(n)->(m)` | matrix-vector multiplication
matmul | `(m?,n),(n,p?)->(m?,p?)` | combination of the four above
outer_inner | `(i,t),(j,t)->(i,j)` | inner over the last dimension, outer over the second to last, and loop/broadcast over the rest.
cross1d | `(3),(3)->(3)` | cross product where the last dimension is frozen and must be 3

This syntax is used by numpy, dask, numba, and likely other parts of the pydata ecosystem do describe gufuncs.

### New idea: grid_ufunc

I think we should refactor all of xgcm's primitive operations to be gufuncs. However, we need an additional layer of complexity, because xgcm deals not just with raw dimensions but with _axes_. We could imagine generalizing the gufunc signature as follows.

name | signature
-- | -- 
diff_center_to_left(a) | `(X:center)->(X:left)`
interp_center_to_left(a) | `(X:center)->(X:left)`
diff_left_to_center(a) | `(X:left)->(Y:center)`
interp_left_to_center(a) | `(X:left)->(Y:center)`

Here `X` and `Y` are arbitrary axis names, while `center` and `left` are xgcm [axis positions](https://xgcm.readthedocs.io/en/latest/grids.html#axes-and-positions). These would be translated to core dimensions, and the low-level numerical routine eventually called via either `xr.apply_ufunc` or else with an xgcm-specific wrapper layer.

Implementing one of these functions might look like

```python
@grid_ufunc(signature="(X:center)->(X:left)")
def diff_center_to_left(a):
    return a - np.roll(a, -1)
```

Here the decorator would be taking care of all the coordinate / dimension wrapping and eventually calling the low-level numerical routine.

For the existing core xgcm routines (e.g. diff, interp), this would be overkill. But it would allow us to start implementing more complex operations. For example
```python
@grid_ufunc(signature="(X:left,Y:center),(X:center,Y:left)->(X:left,Y:left)")
def cgrid_vorticity(u, v):
    # do complicated stuff
    return vort
```
(We could also add metrics to the function and signature.)

If this existed, all of the [laplacian kernels](https://github.com/ocean-eddy-cpt/gcm-filters/blob/master/gcm_filters/kernels.py) from gcm-filters could become xgcm grid_ufuncs (cc @noraloose).

- 2D scalar Laplacian signature: `(X:center,Y:center)->(X:center,Y:center)`
- 2D c-grid vector Laplacian signature: `(X:left,Y:center),(X:center,Y:left)->(X:left,Y:center),(X:center,Y:left)`

Thoughts?

(@tomnicholas)

### PR DESCRIPTION
**Pull request by TomNicholas**
The **big** PR to implement the internal refactoring suggested by @rabernat in #344. After this we won't need a parallel `grid_ufunc_refactor_project` branch any more.

Means that `Grid` no longer uses `neighbour_binary_func` at all (so `_grid_func` has been deleted). However `Axis` still does use  `neighbour_binary_func` - this will be removed in the next version (via #405).

 - [x] Closes #344
 - [x] Tests added (the whole of `test_grid_ufunc.py`)
 - [x] Passes `pre-commit run --all-files`
 - [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
 - [x] New functions/methods are listed in `api.rst`

### CODE REVIEW

**Comment by TomNicholas** [comment]
Something is going wrong with the pre-commit on the repo - it all passes locally for me, but on the CI mypy fails with

```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

xgcm/padding.py:395: error: Item "None" of "Optional[Dict[str, Tuple[int, int]]]" has no attribute "values"
xgcm/grid_ufunc.py:413: error: Unexpected keyword argument "include_extras" for "get_type_hints"
/home/runner/.cache/pre-commit/repoyno4hcj9/py_env-python3.8/lib/python3.8/site-packages/mypy/typeshed/stdlib/typing.pyi:633: note: "get_type_hints" defined here
xgcm/test/test_grid_ufunc.py:2: error: Module "typing" has no attribute "Annotated"
Found 3 errors in 3 files (checked 24 source files)
```
(which shouldn't be an error for python 3.9+), and black fails with all sorts of tiny and pointless changes.

**Comment by TomNicholas** [comment]
> Something is going wrong with the pre-commit on the repo

mypy error fixed by #480, but the black error is still there

File: xgcm/test/test_metrics_ops.py [hunk 1]
```diff
@@ -6,8 +6,21 @@
 from xgcm.grid import Grid
 from xgcm.test.datasets import datasets_grid_metric
 
+GRID_UFUNC_REFACTOR_ERR = (
+    "Metrics have not yet been implemented for the grid ufunc refactor"
+)
+
```
Comment (jbusecke) [inline xgcm/test/test_metrics_ops.py:12]:
```suggestion
```

Comment (jbusecke) [reply xgcm/test/test_metrics_ops.py:12]: This does not seem to be used anywhere?

**Review by jbusecke — COMMENTED** [review]
Couple of small things to check. Otherwise I am ok with this! Cannot wait to merge this.

File: xgcm/grid.py [hunk 2]
```diff
@@ -2263,42 +2586,53 @@ def transform(self, da, axis, target, **kwargs):
...
-        ----------
-        vector : dict
-            A dictionary with two entries. Keys are axis names, values are
-            vector components along each axis.
-        to : {'center', 'left', 'right', 'inner', 'outer'}
-            The direction in which to shift the array. If not specified,
-            default will be used.
-        boundary : {None, 'fill', 'extend'}
-            A flag indicating how to handle boundaries:
+def _select_grid_ufunc(funcname, signature: _GridUFuncSignature, module, **kwargs):
```
Comment (rabernat) [inline xgcm/grid.py:2590]: Just noting that this function is a gorgeous little piece of python.

**Review by rabernat — APPROVED** [review]
Looking at everything all together in one PR makes me realize what a staggering piece of work this has been. 👏  Tom and Julius for pushing this through.

The overall code architecture is brilliant. The tests are incredibly comprehensive. The docs are detailed and useful. I honestly looked for nits to pick and didn't find any.

🚀

File: setup.py [hunk 3]
```diff
@@ -28,7 +28,7 @@
 extras_require["dev"] = extras_require["complete"] + [
     "pytest",
     "pytest-cov",
-    "flake8",
+    "scipy" "flake8",
```
[the remark this answers is not in the document]
Comment (TomNicholas) [reply setup.py:31]:
```suggestion
    "scipy",
    "flake8",
```

**Pull request merged**